bob-backend/src/pearl/disk_controller.rs

```diff
@@ -561,30 +572,35 @@ impl DiskController {
         force_delete: bool,
     ) -> Result<u64, Error> {
         if *self.state.read().await == GroupsState::Ready {
-            let vdisk_group =
-                if !force_delete {
-                    match self.find_group(&op).await {
-                        Ok(group) => group,
-                        Err(_) => return Ok(0)
-                    }
-                } else {
-                    // we should create group only when force_delete == true
-                    match self.get_or_create_pearl(&op).await {
-                        Ok(group) => group,
-                        Err(err) => {
-                            error!(
-                                "DELETE[alien][{}] Cannot find group, op: {:?}, err: {}",
-                                key, op, err
-                            );
-                            return Err(Error::vdisk_not_found(op.vdisk_id()));
-                        }
+            let mut result = 0;
+            let mut err = None;
+            let groups = self.find_all_groups(&op).await;
+            if force_delete && groups.is_empty() {
+                // If delete is forced we need to create at least one group
+                match self.get_or_create_pearl(&op).await {
+                    Ok(group) => match group.delete(key, meta, StartTimestampConfig::new(false), force_delete).await {
+                        Ok(r) => result += r,
+                        Err(e) => err = Some(self.process_error(e).await),
+                    },
+                    Err(e) => {
+                        error!(
+                            "DELETE[alien][{}] Cannot find group, op: {:?}, err: {}",
+                            key, op, e
+                        );
+                        err = Some(Error::vdisk_not_found(op.vdisk_id()))
                     }
-                };
-
-            match vdisk_group.delete(key, meta, StartTimestampConfig::new(false), force_delete).await
-            {
-                Err(e) => Err(self.process_error(e).await),
-                Ok(x) => Ok(x),
+                }
+            }
```

Comment:
You can safely add `else` here because `if` statement has `groups.is_empty()` condition, so the loop below does nothing in this case (the collection is empty). If you do this, then the code above can be simplifying by returning errors immediately (no need to store them in `err` variable)